File: web/rgbd-js/color_decoder.js
import { NativeYuvFrame, YuvFrame } from "./yuv_frame.js";

export const RGBD_COLOR_CODEC_TYPE_VP8 = 0;

export class NativeColorDecoder {
  constructor(wasmModule, colorCodecType) {
    this.wasmModule = wasmModule;
    this.ptr = this.wasmModule.ccall('rgbd_color_decoder_ctor', "number", ["number"], [colorCodecType]);
  }

  close() {
    this.wasmModule.ccall("rgbd_color_decoder_dtor", null, ["number"], [this.ptr]);
  }

  decode(colorBytes) {
    const colorBytesPtr = this.wasmModule._malloc(colorBytes.byteLength);
    this.wasmModule.HEAPU8.set(colorBytes, colorBytesPtr);
    const yuvFramePtr = this.wasmModule.ccall("rgbd_color_decoder_decode",
                                              "number",
                                              ["number", "number", "number"],
                                              [this.ptr, colorBytesPtr, colorBytes.byteLength]);
    this.wasmModule._free(colorBytesPtr);

    const nativeYuvFrame = new NativeYuvFrame(this.wasmModule, yuvFramePtr);
    const yuvFrame = new YuvFrame(nativeYuvFrame);
    nativeYuvFrame.close();
    return yuvFrame;
  }
}

File: python/pyrgbd/color_decoder.py
from enum import IntEnum
from ._librgbd_ffi import lib
from .yuv_frame import NativeYuvFrame, YuvFrame
from .utils import cast_np_array_to_pointer
import numpy as np


class ColorCodecType(IntEnum):
    VP8 = lib.RGBD_COLOR_CODEC_TYPE_VP8


class NativeColorDecoder:
    def __init__(self, color_codec_type: ColorCodecType):
        # Setting lib.VP8 assuming since it is the only codec for now.
        # Fix this later when a codec gets added.
        self.ptr = lib.rgbd_color_decoder_ctor(color_codec_type)

    def close(self):
        lib.rgbd_color_decoder_dtor(self.ptr)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    def decode(self, color_frame_bytes: np.ndarray) -> YuvFrame:
        native_yuv_frame_ptr = lib.rgbd_color_decoder_decode(
            self.ptr,
            cast_np_array_to_pointer(color_frame_bytes),
            color_frame_bytes.size,
        )
        with NativeYuvFrame(native_yuv_frame_ptr) as native_yuv_frame:
            yuv_frame = YuvFrame.from_native(native_yuv_frame)
        return yuv_frame
